nearlib/nearclient.js

```diff
@@ -8,39 +8,85 @@ function _arrayBufferToBase64(buffer) {
-            account_id: account_id,
-        });
-        return viewAccountResponse;
+    async viewAccount(accountId) {
+        const response = await this.jsonRpcRequest('abci_query', [`account/${accountId}`, '', '0', false]);
+        return JSON.parse(_base64ToBuffer(response.response.value).toString());
     }
 
-    async submitTransaction (signedTransaction) {
+    async submitTransaction(signedTransaction) {
         const buffer = SignedTransaction.encode(signedTransaction).finish();
         const transaction = _arrayBufferToBase64(buffer);
-        const data = { transaction };
-        try {
-            return await this.request('submit_transaction', data);
-        } catch(e) {
-            if (e.response) { 
-                throw new Error(e.response.text);
-            }
-            throw e;
+        const params = [transaction];
+        const response = await this.jsonRpcRequest('broadcast_tx_async', params);
+        response.hash = Buffer.from(response.hash, 'hex');
+        return response;
+    }
+
+    async callViewFunction(contractAccountId, methodName, args) {
+        if (!args) {
+            args = {};
+        }
+        const serializedArgs = Buffer.from(JSON.stringify(args)).toString('hex');
+        const result = await this.jsonRpcRequest('abci_query', [`call/${contractAccountId}/${methodName}`, serializedArgs, '0', false]);
+        const response = result.response;
+        const logs = response.log || '';
+        logs.split("\n").forEach(line => {
+            console.log(`[${contractAccountId}]: ${line}`);
+        });
+        // If error, raise exception after printing logs.
+        if (response.code != 0) {
+            throw Error(response.info)
         }
+        const json = JSON.parse(_base64ToBuffer(response.value).toString());
+        return json;
     }
 
-    async getNonce (account_id) {
-        return (await this.viewAccount(account_id)).nonce + 1;
+    async getTransactionStatus(transactionHash) {
+        const encodedHash = _arrayBufferToBase64(transactionHash);
+        const response = await this.jsonRpcRequest('tx', [encodedHash, false]);
+        // tx_result has default values: code = 0, logs: '', data: ''.
+        let status;
+        const code = response.tx_result.code || 0;
+        switch (code) {
+            case 0: status = 'Completed'; break;
+            case 1: status = 'Failed'; break;
+            case 2: status = 'Started'; break;
+            default: status = 'Unknown';
+        }
+        const logs = response.tx_result.log || '';
+        return {logs: logs.split('\n'), status, value: response.tx_result.data };
+    }
+
+    async getNonce(accountId) {
+        return (await this.viewAccount(accountId)).nonce + 1;
+    }
+
+    async jsonRpcRequest(method, params) {
+        const request = {
+            jsonrpc: '2.0',
+            method,
+            params,
+            id: 'dontcare',
```

Comment:
are you sure there's not gonna be collision of ids? Like maybe it matters when you issue multiple requests at the same time?

Reply:
made it a timestamp